### ISSUE TITLE
Collapsed toots cause weird scrolling

### ISSUE THREAD

**Issue by Felthry** [glitch-soc/mastodon#417]
When an auto-collapsed toot makes its way into your home timeline (and probably others as well), if that timeline isn't scrolled all the way to the top (I habitually scroll down a bit so that new notifications don't automatically scroll the old ones down, and I can keep my place), it appears that the interface calculates the new scroll offset based on the uncollapsed toot, so that instead of the user's view being unmoving as intended, the timeline scrolls itself down whenever someone posts a new toot long enough to autocollapse.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by ThibG** [comment]
This is because auto-collapsed toots are rendered expanded once before getting collapsed (the decision to auto-collapse is taken in `componentDidMount`, because it uses `node.clientHeight`.
On the other hand, `ScrollableList` only updates the position when the number of children increases, so the second rendering of auto-collapsed toots will not adjust the scroll position…

I'm not too sure how to deal with that.

**Comment by marrus-sh** [comment]
There *used* to be a function (`onHeightChange` i believe) that you could call from inside the component to recalculate the height whenever the height of the component changed—this isn't just a glitchsoc problem, but also an upstream one when opening/closing CWs. I am not sure if that function still exists or if there is a different method now, but my guess is that whatever it is, it is not being called on the *initial* collapse and only for ones thereafter. (That, or, we have broken it somehow.)

**Comment by Felthry** [comment]
Is it possible to move the auto-collapse decision to be earlier in the queue? Or possibly add some scrolling commands inside the collapse function?

**Comment by ThibG** [comment]
I'm afraid I can't see how to move the auto-collapse decision earlier, unfortunately.
From what I understand, `componentDidMount` is the earliest we can make that decision if it is to be based on the size of the rendered status. `componentDidMount` is called immediately after the status has been `render()`ed and will trigger a re-render, but `ScrollableList`'s `componentDidUpdate` is called after the first render, not the second. Subsequent renders of the statuses won't cause the `ScrollableList` to render again, which is expected, but this means the scroll code doesn't get called either.

Basically, it's idiomatic to update the scroll position in the parent's `componentDidMount` when adding an item, but it assumes items won't change sizes once added.

I can't think of another way than issuing a second scrolling command after the second render, but doing so is very tricky (you only want to do that after an automatic-collapse, which isn't easy to express).

I think that this issue will also occur, albeit less dramatically, with my PR #426.
